packages/foam-vscode/src/features/wikilink-provider.ts

```diff
@@ -0,0 +1,183 @@
+    private getNavigationMode: () => string
+  ) {}
+
+  public async provideDefinition(
+    document: vscode.TextDocument,
+    position: vscode.Position,
+    token: vscode.CancellationToken
+  ): Promise<vscode.LocationLink[] | vscode.Definition> {
+    const navMode = this.getNavigationMode();
+    if (navMode === OPEN_LINK_NAV_MODE || navMode === NAVIGATION_DISABLED) {
+      return;
+    }
+
+    const resource = this.parser.parse(document.uri, document.getText());
+    const targetLink: ResourceLink | undefined = resource.links.find(link =>
+      Range.containsPosition(link.range, {
+        line: position.line,
+        character: position.character,
+      })
+    );
+
+    if (!targetLink) {
+      return;
+    }
+
+    const uri = this.workspace.resolveLink(resource, targetLink);
+    if (URI.isPlaceholder(uri)) {
+      if (
+        navMode === MIXED_NAV_MODE ||
+        vscode.workspace.workspaceFolders.length < 1
+      ) {
+        return;
+      }
+
+      const newNoteUri = toVsCodeUri(
+        URI.createResourceUriFromPlaceholder(
+          vscode.workspace.workspaceFolders[0].uri,
+          uri
+        )
+      );
+      const newNoteTitle = uri.path.split('/').slice(-1)[0];
+      await createNoteForPlaceholderWikilink(newNoteTitle, newNoteUri);
+    }
+
+    const targetResource = this.workspace.get(uri);
+    const targetRange = findFirstLineOfContent(targetResource.source.text);
+
+    const result: vscode.LocationLink = {
+      originSelectionRange: toVsCodeRange(targetLink.range),
+      targetUri: toVsCodeUri(uri),
+      targetRange: null, // Disable the preview feature to avoid duplication with the over feature.
+      targetSelectionRange: toVsCodeRange(targetRange),
+    };
+    return [result];
+  }
+
+  public provideDocumentLinks(
+    document: vscode.TextDocument
+  ): vscode.DocumentLink[] {
+    const navMode = this.getNavigationMode();
+    if (
+      navMode === GO_TO_DEFINITION_NAV_MODE ||
+      navMode === NAVIGATION_DISABLED
+    ) {
+      return;
+    }
+
+    const resource = this.parser.parse(document.uri, document.getText());
+
+    const targets: { link: ResourceLink; target: URI }[] = resource.links.map(
+      link => ({
+        link: link,
+        target: this.workspace.resolveLink(resource, link),
+      })
+    );
+
+    return targets
+      .filter(
+        o => navMode === OPEN_LINK_NAV_MODE || URI.isPlaceholder(o.target)
+      )
+      .map(o => generateDocumentLink(o.link, o.target));
+  }
+}
+
+const generateDocumentLink = (link: ResourceLink, target: URI) => {
+  const command = OPEN_COMMAND.asURI(toVsCodeUri(target));
+  const documentLink = new vscode.DocumentLink(
+    toVsCodeRange(link.range),
+    command
+  );
+  documentLink.tooltip = URI.isPlaceholder(target)
+    ? `Create note for '${target.path}'`
+    : `Go to ${URI.toFsPath(target)}`;
+  return documentLink;
+};
+
+export const findFirstLineOfContent = (rawContent: string): Range => {
```

Comment:
Is it OK to implement a custom parsing algorithm ? In the long term, it's maybe not the ideal place to stay.

I tried to use `targetResource.source.contentStart` but it's not sufficient, and it skips systematically two lines after the front matter if there is one (not sure if it's the intended behavior).

Reply:
Feels like a bug with `contentStart` (but I wouldn't fix it in this PR).

Reply:
I agree, this PR it's not the place to fix it.